KEP 1645: add labels and annotations export (#4922)
kubernetes/enhancements: keps/sig-multicluster/1645-multi-cluster-services-api/README.md

        // +optional
        metav1.ObjectMeta `json:"metadata,omitempty"`
        // +optional
        Spec ServiceExportSpec `json:"spec,omitempty"`
        // +optional
        Status ServiceExportStatus `json:"status,omitempty"`
}

// ServiceExportSpec describes an exported service and extra exported information
type ServiceExportSpec struct {
        // +optional

>         ExportedLabels map[string]string `json:"exportedLabels"`

Author: Ah yep nice catch! It's already like that in the associated PR for the CRD but the KEP should be amended indeed

>         ExportedLabels map[string]string `json:"exportedLabels"`

Reviewer: could you please explain which case should use exportedLabels/exportedAnnotations? provide an example?

it will be really helpful to provide some guidance on what information should be in the annotation/labels on the serviceExport or exportedLabels/exportedAnnotations on the spec of the serviceExport.

Author: This is to add any labels or annotations the user wants to the ServiceImport (+ the associated derived service depending if the MCS-API implementation uses that). The use case can vary a lot depending of the context and the environments.

For instance, if there is an operator selecting ServiceImport/Service based on label selector you would need some kind of labels. Prometheus operator is one example of that, it has a service selector on the its ServiceImport resource and if the implementation uses a derived service you could supposedly add labels through that and uses a ServiceMonitor to monitor a service deployed on remote clusters too.

For annotations, it also depends a lot, that would be probably things that would change the behavior of Service or ServiceImport based on some annotation. As an example Cilium have many annotation that updates the Service behavior, here is a couple example that I know OTOH: https://docs.cilium.io/en/stable/network/servicemesh/envoy-load-balancing/#add-proxy-load-balancing-annotations-to-the-services  & https://docs.cilium.io/en/latest/network/clustermesh/services/#synchronizing-kubernetes-endpointslice-beta, and there is even the topology mode for a regular installation with kube-proxy although that is replaced by the dedicated "trafficDistribution" field but still.

Reviewer: I see, these labels/annotations will be added in the derived-service, so it's not the label/annotation for the service or serviceExport itself. 

Could you please explain how these exportedLabels/exportedAnnotations can be used in the serviceImport?

based on the serviceImport definition, `"servcieImport will be introduced to act as the in-cluster representation of a multi-cluster service in each importing cluster."`, exportedLabels/exportedAnnotations is per cluster, how can we aggregate these information on the label of serviceImport?

Reviewer: My concern is, the field name itself reveals some implementation details, especially label/annotation. if i understand it correctly, they're extra information about how to export the service itself. Should we use some generic name? eg, "properties"

Author: > Could you please explain how these exportedLabels/exportedAnnotations can be used in the serviceImport?

The KEP currently states this:
```
The derived service's labels and annotations will be decided
according to the conflict resolution if the set of name/value pairs are not identical
between the constituent clusters.
```

Basically if an implementation chose to add supports for this they will have to make sure all the maps of exported{Annotations,Labels} are identical or take only the oldest ServiceExport maps and raise a conflict like we do for all the other properties. Those will then be transposed to each ServiceImport .metadata.{annotations,labels} and possibly its derived service(s) if the implementation has that.

> My concern is, the field name itself reveals some implementation details, especially label/annotation. if i understand it correctly, they're extra information about how to export the service itself. Should we use some generic name? eg, "properties"

Labels/annotations is a generic thing present on every kubernetes objects, so I am not sure I understand your point unfortunately. This is only about having the possibility of getting some labels and annotations up to the ServiceImport resources.

Reviewer: Thank you for the information. Please correct me if my understanding is wrong.

```
The derived service's labels and annotations will be decided
according to the conflict resolution if the set of name/value pairs are not identical
between the constituent clusters.
```
https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md#labels-and-annotations if i understand it correctly, it's mainly about how to solve the conflicts for the labels/annotations on the original Service itself, not about the new fields in the ServiceExport.

About the labels/annotations, from this [comment ](https://github.com/kubernetes/enhancements/pull/4922#discussion_r1891921225), all the examples given are related to the implementation and then decide where these extra information should be put, either annotation or label. For example, Prometheus operator wants the label from the derived service, and as a result, the information should be in the label.

Author: > if i understand it correctly, it's mainly about how to solve the conflicts for the labels/annotations on the original Service itself, not about the new fields in the ServiceExport.

No the labels/annotations of the original services should be ignored during the export and should not influence the labels/annotations of the ServiceImport (or related derived service)

> About the labels/annotations, from this comment , all the examples given are related to the implementation and then decide where these extra information should be put, either annotation or label. For example, Prometheus operator wants the label from the derived service, and as a result, the information should be in the label.

I am not sure if this is your point but there is no known usage of labels/annotations on the ServiceImport currently which is mainly why this is an optional feature as it's not very useful right now for implementation not using directed services.

Reviewer: `No the labels/annotations of the original services should be ignored during the export and should not influence the labels/annotations of the ServiceImport (or related derived service)` then, need to clarify it in the KEP. There are three types of labels/annotations now: on service itself, on serviceExport itself, servcieExport fields.

`I am not sure if this is your point but there is no known usage of labels/annotations on the ServiceImport resources currently which is mainly why this is an optional feature as it's not very useful right now for implementation not using derived services.`

My concern is more about naming. I would like to put extra information to the servceExport itself. But about where these information will be used, it could be one of label, annotation or even a field in a CR, which depends on the implementation.

Author: We explicitly state this here:

> This does not apply for labels and
annotations which are stored in `ServiceExport` directly in `spec.exportedLabels` and `spec.exportedAnnotations`. Exporting labels and annotations is optionally supported by MCS-API implementations. If supported, annotations or labels must not be exported from the `metadata` of the `Service` or `ServiceExport` resources.

And also we are talking about this in the alternative section as well.

> My concern is more about naming. I would like to put extra information to the servceExport itself. But about where these information will be used, it could be one of label, annotation or even a field in a CR, which depends on the implementation.

Well this not a a huge amount of context, but I guess you could add stuff in annotations yes or propose some new fields depending on what is it.

Reviewer 2: > No the labels/annotations of the original services should be ignored during the export and should not influence the labels/annotations of the ServiceImport (or related derived service)
> 

It seems that https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md#labels-and-annotations was talking about the original service? Does the new spec change the original statement? I wonder what happens if the export and the original service both have a label "x"?

Author: > It seems that https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md#labels-and-annotations was talking about the original service? Does the new spec change the original statement?

AFAIU this is talking about the derived service and was added together with the new fields btw.

> I wonder what happens if the export and the original service both have a label "x"?

Original Service labels are meant to be completely ignored to not have to skip some of the labels/annotation as we would need to figure out a filtering approach to skip some of the annotations/labels in this case.

        // +optional

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: You mean like that we would have some filters instead of the actual values? If so yes that might be one alternative approach that would work too, let's see what other think!

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 3: If the ServiceExport stores the values, we need to describe how updates to those values are handled. I don’t think we’d ever want different values on a Service and its ServiceExport, so having only the names of the labels/annotations is sufficient IMO and avoids some complexity.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: Hmm maybe it's not clear but what I wrote in the KEP is not supposed to look at the corresponding Service at all, like all the labels/annotations exported would be in the ServiceExport spec a bit like what's done in gateway API (which makes me realize that I haven't posted the link to gateway API doc in the description: https://gateway-api.sigs.k8s.io/geps/gep-1762/#labels-and-annotations)

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: Although I was originally in favor of not adding any fields to the ServiceExport object at all, but with @mikemorris suggestion in the meeting I was a bit inclined to do it a bit like GatewayAPI and tried to advocate for the reasons we would like to do that in this PR although a different approach is definitely doable!

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 3: Ah, right — I interpreted the mention of Service in the section title, “Exporting labels/annotations from the Service/ServiceExport objects”, incorrectly! And in fact you do exclude this interpretation further down, “Implementations must not export the annotations or
labels from the metadata of the `Service` or `ServiceExport` resources.”

In that case it would be interesting to consider whether exported metadata is likely to be identical to the existing ServiceExport metadata, or have different values — storing values separately ensures that we can have different values in the ServiceExport and (generated) ServiceImport metadata, whereas a filter approach prevents that entirely.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: If we go on the filtering route there might be annotations that only make sense from a clustermesh POV. This is very Cilium specific but for instance we do have an annotation to prefer the local cluster or the remote cluster (this https://docs.cilium.io/en/latest/network/clustermesh/affinity/#enabling-global-service-affinity). So while it would work on the local (not derived) Service it would makes it a bit weird.

Also you might want to possibly have a different set of labels for your ServiceImport (and derived Service) from your local service. For instance, if you want to select only the derived Service and not potentially the local Service it might be a bit problematic to take the labels (even filtered) from the local Service. We can totally solve that by saying something along those lines in the KEP "if the implementation use a derived Service you must add the label `multicluster.kubernetes.io/derived-service: "true"` " which would be sufficient IMO.

No big blockers though just throwing additional constraints/requirements that we may consider...

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 4: > We can totally solve that by saying something along those lines in the KEP "if the implementation use a derived Service you must add the label `multicluster.kubernetes.io/derived-service: "true"` which would be sufficient IMO.

We have a [similar requirement](https://gateway-api.sigs.k8s.io/geps/gep-1762/#automated-deployments) over in Gateway API for controllers to add a `gateway.networking.k8s.io/gateway-name: my-gateway` label to any generated resources, which we then depend on for testing whether any configured infrastructure labels/annotations have been set on generated resources - I would suggest naming this slightly more generic like `multicluster.kubernetes.io/derived-resource` but I think the concept is sound. I'm not sure if an explicit link to _which_ ServiceImport/Export it's derived from is necessary or if that could be inferred from the name, but it may be helpful.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 4: I would advise against the copy-filter pattern, we tried that on a Gateway API implementation I worked on previously, and it was [a bit burdensome and a source of confusion for users](https://github.com/hashicorp/consul-api-gateway/issues/601).

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 5: The semantics depends on the use case we're trying to solve. My interpretation was that the intent was to transfer verbatim certain labels/annotations values from the exported `Service` to the derived `ServiceImport` (and derived `Service` if the implementation creates one), in which case we'd want only the names specified. But if the intent is to populate derived labels/annotations solely from the `ServiceExport`, then, of course, we'd need the values specified as well. With that approach, if the user wants the derived resource to reflect the exported `Service`'s values, then they would have to know that info when creating the `ServiceExport` and also update the `ServiceExport` accordingly on changes, if necessary. A bit more of a burden as @skitt noted.

I'm not clear on the exact use case for Cilium (Submariner doesn't have a use case for this). @MrFreezeex re: "_you might want to possibly have a different set of labels for your ServiceImport (and derived Service) from your local service_", is this part of Cilium's use case or is this just hypothetical?

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: > I'm not clear on the exact use case for Cilium (Submariner doesn't have a use case for this). @MrFreezeex re: "you might want to possibly have a different set of labels for your ServiceImport (and derived Service) from your local service", is this part of Cilium's use case or is this just hypothetical?

For labels the use case would be beneficial for many implementations I think. This is mostly for third party controllers that have selectors. Regarding ServiceImport specifically I don't think there is any use case so far (but there might be in the future!) but for the derived Service if you can add some labels any controllers that have a service selector would be able to make use of this. For instance, in the ServiceMonitor resource of Prometheus operator you have a service selector which means that if you start populating labels to your derived Service you can theoretically (if you sync EndpointSlice) do some multicluster scraping. If you didn't have labels this would be much harder to integrate with those type of controllers...

About annotations, we have many annotations in Cilium that change the behavior of a Service which may or may not be related directly to the clustermesh implementation (for instance there's an annotation to load balance multiple HTTP2 stream to different backends instead of load balancing at the connection levels like you would typically do).

IMO the best choice for labels and annotations sync would be either:
- Take labels and annotations it from the metadata Service directly. I am not entirely convinced right now about a label/annotations filters per object so I would probably favor only an annotation prefix filter at the controller level (that can probably be configured when installing the software implementing MCS-API, but the KEP should stay vague about that anyway).
- Like the current version this PR is describing: have the labels/annotations in the spec of ServiceExport that are exported verbatim without the local Service involved. This allows to fully decouple labels and annotations of the final ServiceImport (and then the derived Service for most implementation) which is a good thing for labels (I think?) but for annotation could be both nice and weird. The nice part is that you can put specific clustermesh annotation there (which removes the awkwardness of adding to a regular Service which in isolation has nothing to do with clustermesh/multi cluster). The bad part is that you can have annotations that modify the behavior of the Service entirely (such as the topology aware annotation if we stay in the standard annotations) which would end up only in the ServiceImport and its derived Service while not present on the local Service which is kind of arguable because it's not like any of the other fields...

There is also a middle ground/tweak to the first option, which is to merge the annotations (and potentially labels) of both the ServiceExport and the local Service before actually "exporting" it but I am worried in term of UX/coherence here as it's not like any other fields and not obvious...

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: Argh actually I was checking about labels and for instance ArgoCD use a label to track the Application (https://argo-cd.readthedocs.io/en/latest/user-guide/annotations-and-labels/) so unlike what I just said above there is definitely a need for labels filtering too (at least to make sure that they don't conflict)... So what you were suggesting @tpantelis @skitt looks like a solid option too.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 5: There’s probably use cases for both approaches. This PR describes copying from the ServiceExport which is fine. The filter approach could achieved via key prefixes, either well known or configured, or specific keys. Perhaps this should be left up to implementations to define and implement as they see fit and either leave it out of the KEP or perhaps mention it as optional and not define specifics.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: > Perhaps this should be left up to implementations to define and implement as they see fit and either leave it out of the KEP or perhaps mention it as optional (in the Alternatives section) but not define specifics.

WDYM? Because like if we add some fields to the ServiceExport implementation must at least make some use of those fields (whether it's the field to filter like you are proposing or directly the map like the current version of this PR).

Are you suggesting that the new fields would do something (TBD what exactly) and outside of that the implementation may have some well known prefix somehow built-in that he could take from the local Service directly for instance? Meaning that in the KEP we would describe what those fields do without enforcing that it is the only way that an implementation may use to export labels and annotations?

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 5: So we've been discussing two potential ways to propagate labels/annotations to derived resource(s) with potential use cases for each.

1) Specify the exact keys and values to copy in the `ServiceExport` spec.

2) Specify the keys that are copied from the `Service` instance. 

Both approaches have pros and cons depending on the use case. So should we pick one or the other to formally define/describe in the KEP and possibly discuss the other in the _Alternatives_ section? Or should we define both to give users ultimate flexibility on how and from where to propagate labels/annotations?

Whichever is formally defined, will it be required or optional for implementations to support? This seems geared towards supporting a derived `Service`, which is not required by the spec. Submariner does  not create derived `Services` and would have no use to propagate labels/annotations to the `ServiceImport`.

Given that, should we then not formally define any way to propagate labels/annotations and let implementations handle it (or not) as they see fit? Both approaches could be implemented via key prefixes without adding constructs to the `ServiceExport` CRD.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 5: > I think we could make it optional though

yes I think it should be optional given that the motivation is to support derived `Services` which are implementation-specific and not required.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: I pushed some new changes to clearly mark labels and annotation export as optional and detailed more about the filtering alternative, let me know what do you think about this new version :pray:.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 6: > * Specify the exact keys and values to copy in the `ServiceExport` spec.
> 
> * Specify the keys that are copied from the `Service` instance.

How about a CEL expression that takes a key and returns true or false? It's expensive to evaluate but the set of keys is typically stable so the filter shouldn't be recomputed often.

Potentially two expressions: one for labels, one for annotations

Why is this good?
- flexible
- lets you do things like wildcards but with an exception
- very configurable handling for unrecognized labels / annotations
- allows for rules like "export either both of these labels or neither of them" or "export label A or B, but never both"
- extensible (I think) to filtering on values as well

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 6: AIUI server-side apply lets us go further and decide whether to become a field manager for a particular key, and if so to:
- take over management
- jointly manage the key
- set / update the value but only if there is no other declared manager
- skip this key and not copy it to the target

A CEL expression could return which of those four options to pick, per label / annotation key.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: > AIUI server-side apply lets us go further and decide whether to become a field manager for a particular key

I don't think this is a case where server-side apply would be relevant here because this is not a conflict problem within the same cluster between different actors.

The context is that we need to figure out a set of labels/annotations that are getting exported from each "local Services" in a way that would reasonably not result in a "cross-cluster conflict"  even if third party tools are used that might add extra labels/annotations (kubectl with the last applied annotation, argocd with the application labels are example of that).

> How about a CEL expression that takes a key and returns true or false? It's expensive to evaluate but the set of keys is typically stable so the filter shouldn't be recomputed often.

This is quite an interesting idea to apply CEL here. I see it has a "copy-filter" approach but in a much more flexible way. But it's also way more complex I am not sure if as a user of an MCS-API implementation, I would want to write some CEL expressions to filter my labels/annotations and also probably making sure that this expression would be the same on all similar ServiceExport across clusters...

So IMO this has the same benefits of "copy-filter" approach (for instance a list of prefix) which is consistency with the rest of the API: the source of truth is on the Service itself with more flexibility but also more complex for users.

I am not entirely sure if this is really better than the the way this KEP is currently proposing which is trying to skip this copy-filter problem entirely by adding dedicated fields in the spec that won't be touched by third party tooling which may result in a conflict when reconciling those cross clusters.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 6: @MrFreezeex I'd like the CEL expression idea to appear in the list of alternatives; it doesn't have to get listed as the winning choice.

What I disagreed with, lightly, was the implication that we had to choose between exactly 2 design options here. Actually there are more.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Author: Well we never considered CEL but AFAICT it's still is fundamentally a "copy-filter" approach which we discussed here and with some points against it in `Exporting labels/annotations from the Service/ServiceExport objects`. The only thing that might not be relevant for the CEL proposal is the last point about flexibility which is traded by more complexity. Will try to rephrase at to that last point to include writing CEL expressions.

>         ExportedAnnotations map[string]string `json:"exportedAnnotations"`

Reviewer 6: (OK, this is resolved)

ensure that a name is shared by multiple services within the namespace if and
only if they are instances of the same service.

Most information about the service, including ports, backends, topology and

> session affinity, will continue to be stored in the `Service` objects, which

Author: This is in the ServiceExport section so I think it's (mostly?) clear that we are talking about `ServiceExport` and the `Service` of the same name. I can definitely change something though if you have a suggestion/hints though

> session affinity, will continue to be stored in the `Service` objects, which

Reviewer 5: OK, you're right - this is solely about `ServiceExport`.

ensure that a name is shared by multiple services within the namespace if and
only if they are instances of the same service.

Most information about the service, including ports, backends, topology and
session affinity, will continue to be stored in the `Service` objects, which
are each name mapped to a `ServiceExport`. This does not apply for labels and
annotations which are stored in `ServiceExport` directly in `spec.exportedLabels`
and `spec.exportedAnnotations`. Exporting labels and annotations is optionally
supported by MCS-API implementations. If supported, annotations or labels must

> not be exported from the `metadata` of the `Service` or `ServiceExport` resources.

Author: I am not sure I agree with this because it is intentionally vague about what implementation should do (does precedence means merging or completely overriding?), if we find the need to change that behavior I would just edit those sentence rather than having something somewhat unclear to begin with

service's session affinity will be decided according to the conflict resolution
policy.

#### Labels and annotations

Labels and Services affects a service as a whole for a given consumer. The derived
service's labels and annotations will be decided according to the conflict resolution

> if the maps are not identical.

Author: In the current form I would be in favor of explicitly stating that the maps should be identical (and not the union or things like that) WDYT?

> if the maps are not identical.

Reviewer 5: Ok - makes sense. Instead of the term "map" which is more of a programming language construct, how about:

"_The derived service's labels and annotations will be decided according to the conflict resolution policy if the set of name/value pairs are not identical between the constituent clusters._"

service's session affinity will be decided according to the conflict resolution
policy.

#### Labels and Annotations

> If supported, exporting Labels and Annotations would affect a service as a whole

Reviewer 6: (nit)
```suggestion
If supported, exporting labels and annotations would affect a Service as a whole
```

> If supported, exporting labels and annotations would affect a `Service` as a whole

Reviewer 4: ```suggestion
If supported, exporting labels and annotations would affect a service as a whole
```
For consistency with the above sections which do not explicitly mention the `Service` resource - my reading is that a derived `Service` resource is an optional implementation detail which is one component of the "service as a whole".

Reviewer 4: I see this would revert @sftim's prior suggested change in https://github.com/kubernetes/enhancements/pull/4922#discussion_r1836993024

### Export via annotation

> `ServiceExport` as described has almost no spec and seems like it could just be

Reviewer 3: I would rephrase this to make it historical, since the alternative was considered and rejected when ServiceExport had no spec:

```suggestion
`ServiceExport` initially had no spec and seemed like it could just be
```

### Export via annotation

> `ServiceExport` initially had no spec and seemed like it could just be

Reviewer 4: Out of scope for this change, but I think the original rationale here is actually outdated - the `Service` resource does have a `status.conditions` field now https://kubernetes.io/docs/reference/kubernetes-api/service-resources/service-v1/#ServiceStatus

Author: This is about ServiceExport spec so I would say it should still be right?

Reviewer 4: This section is describing why a `ServiceExport` resource with no `spec` was originally proposed at all - the `ServiceExport` status was intended to fill a UX gap from `Service` lacking a `status.conditions` field to message if an annotation-based export would have been successful or conflicted. It's not directly applicable to the focus of this PR though, I may update the language for historical purposes in a separate PR.

ServiceExport oldness makes resolving conflicts straightforward, and this
alternative conflict resolution algorithm could hinder this ease of use.

### Exporting labels/annotations from the Service/ServiceExport objects

`Service` and `ServiceExport` have labels and annotations which could be used during
export and propagated to the `ServiceImport`. However various tools such as kubectl or
ArgoCD add some labels and annotations which would then need to be actively
filtered to avoid any conflict. Filtering those labels and annotations is not
something easy and we chose to avoid this problem entirely by not using the metadata
object and adding dedicated fields in the spec of the `ServiceExport` resource.

Also if we were using the labels and annotations from the metadata of either the

> `ServiceExport` or `Service` resources, it may be more confusing for users as it
> would be the only fields present in both resources. For instance, should an

Reviewer 4: ```suggestion
`ServiceExport` or `Service` resources, it may be more confusing for users.
For instance, should an
```
I didn't quite understand this, and think the point is clear without this phrase.

implementation merge the labels/annotations from both objects? Should it favor one?
Should it takes only from the `Service` object? With dedicated fields for labels

> and annotations in the spec of the `ServiceExport` resource, it may becomes more
> straightforward that each resource have their own labels and annotations in their
> metadata and that the exported labels and annotations are from the dedicated
> fields in the `ServiceExport` spec.

Reviewer 4: ```suggestion
and annotations in the spec of the `ServiceExport` resource, we expect it will be
straightforward that while each resource has their own labels and annotations in their
metadata, the exported labels and annotations are from the dedicated fields in the
`ServiceExport` spec.
```
Minor phrasing nit.

We also favored dedicated fields on the `ServiceExport` resource to allow for better
flexibility, as it will allow to export labels and annotations fully decorrelated

> from the `Service` and `ServiceExport` metadata. More flexibility could also be
> achieved with CEL expression on the `ServiceExport` at the cost of greater

Reviewer 4: ```suggestion
from the `Service` and `ServiceExport` metadata. 

This implementation does not preclude the possibility of additionally adopting a
"copy/filter" strategy (either as global configuration for a particular implementation, or
granularly as additional `ServiceExport` fields) if the need arises. Such an strategy could
select specific label or annotation keys to sync from the `Service` by exact name, by
prefix, or further flexibility could be achieved with CEL expressions at the cost of
greater
```

Feel free to reject this suggestion - from the long thread above this felt like this possibility could be worth mentioning but perhaps the "`exportedLabels` and `exportedAnnotations` MUST take precedence" snippet I suggested adding above is a sufficent oblique reference without introducing confusion.
